File: q1751/q1751.java
class Solution {
    public int maxValue(int[][] events, int k) {
        Arrays.sort(events, (e1, e2) -> e1[0] - e2[0]);

        int[][] memo = new int[events.length][k + 1];
        for (int[] row : memo)
            Arrays.fill(row, -1);

        return dpHelper(0, events, k, memo);
    }

    private int dpHelper(int currIdx, int[][] events, int k, int[][] memo) {
        if (k == 0 || currIdx == events.length)
            return 0;

        if (memo[currIdx][k] != -1)
            return memo[currIdx][k];

        // attend the current event, and recursively compute from the next possible
        // start time
        int nextCompatibleEventIndex = findNextCompatibleEventIndex(events, events[currIdx][1]);
        int val1 = events[currIdx][2] + dpHelper(nextCompatibleEventIndex, events, k - 1, memo);

        // do not attend the event, and recursively compute from the next immediate
        // start time
        int val2 = dpHelper(currIdx + 1, events, k, memo);

        memo[currIdx][k] = Math.max(val1, val2);
        return memo[currIdx][k];
    }

    private int findNextCompatibleEventIndex(int[][] events, int currEndTime) {
        int l = 0;
        int r = events.length;
        while (l < r) {
            int m = (l + r) / 2;
            if (events[m][0] <= currEndTime) {
                l = m + 1;
            } else {
                r = m;
            }
        }
        return l;
    }
}

File: q1751/q1751.cpp
class Solution
{
public:
    int maxValue(vector<vector<int>> &events, int k)
    {
        sort(events.begin(), events.end());
        vector<vector<int>> memo(events.size(), vector<int>(k + 1, -1));
        return dpHelper(0, events, k, memo);
    }

    int dpHelper(int currIdx, const vector<vector<int>> &events, int k, vector<vector<int>> &memo)
    {
        // if currIdx is out of range or we do not have any k left, return 0
        if (currIdx >= events.size() || k == 0)
            return 0;

        // if we previously have computed the value, return that value
        if (memo[currIdx][k] != -1)
            return memo[currIdx][k];

        // attend the current event, and recursively compute from the next compatible index (start after curr event ends)
        int nextCompatibleEventIdx = findNextCompatibleEventIdx(events, events[currIdx][1]);
        int val1 = events[currIdx][2] + dpHelper(nextCompatibleEventIdx, events, k - 1, memo);

        // do not attend the current event, and recursively compute from the immediate next index
        int val2 = dpHelper(currIdx + 1, events, k, memo);

        memo[currIdx][k] = max(val1, val2);
        return memo[currIdx][k];
    }

    int findNextCompatibleEventIdx(const vector<vector<int>> &events, int currEndTime)
    {
        int l = 0;
        int r = events.size();
        while (l < r)
        {
            int m = (l + r) / 2;
            if (events[m][0] <= currEndTime)
            {
                l = m + 1;
            }
            else
            {
                r = m;
            }
        }
        return l;
    }
};